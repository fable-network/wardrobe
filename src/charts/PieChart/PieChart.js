import React from 'react';
import PropTypes from 'prop-types';
import styled, { withTheme } from 'styled-components';
import { theme as defaultTheme } from '../../theme';
import HighChart from '../HighChart';
import { tooltipCss, hideCreditsCss } from '../HighChartCss';

const getColors = p => {
  const colors = p.colors || p.theme.piePalette;
  let colorsCss = '';
  colors.forEach((color, i) => {
    colorsCss += `
        .highcharts-color-${i} {
          fill: ${color} !important;
          stroke: ${color};
      }`;
  });
  return colorsCss;
};

const Wrapper = styled.div`
  .ft-wardrobe-pie-chart {
    ${getColors};

    ${tooltipCss} ${hideCreditsCss};
  }
`;

class PieChart extends React.PureComponent {
  render() {
    const {
      data,
      colors,
      title,
      tooltip,
      allowUpdate,
      theme = defaultTheme,
      height = null,
    } = this.props;
    if (!data) return null;

    const options = {
      title: title && {
        text: title,
      },
      chart: {
        type: 'pie',
        backgroundColor: theme.white,
        styledMode: true,
        height,
      },
      plotOptions: {
        pie: {
          allowPointSelect: true,
          cursor: 'pointer',
          dataLabels: {
            enabled: false,
          },
        },
      },

      tooltip: {
        enabled: true,
        outside: false, // Don't change unless you want to re-do the whole tooltip styling
        useHTML: true,
        shadow: true,
        backgroundColor: theme.white,
        borderColor: theme.grey04,
        borderRadius: 0,
        padding: 8,
        formatter: function formatter() {
          const body = tooltip({ ...this });
          return `<span class="ft-tooltip">
            <span class="ft-tooltip-header">${this.key}</span>
            <span class="ft-tooltip-body">${body}</span>
          </span>`;
        },
      },
      legend: {
        enabled: false,
      },
      series: [
        {
          data,
        },
      ],
    };
    return (
      <Wrapper colors={colors}>
        <HighChart
          options={options}
          className="ft-wardrobe-pie-chart"
          allowChartUpdate={allowUpdate}
        />
      </Wrapper>
    );
  }
}

PieChart.defaultProps = {
  tooltip: ({ y }) => y,
  allowUpdate: false,
};

PieChart.propTypes = {
  title: PropTypes.string, // optional
  data: PropTypes.arrayOf(
    PropTypes.shape({
      y: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
    })
  ).isRequired,
  tooltip: PropTypes.func, // optional
  colors: PropTypes.arrayOf(PropTypes.string), // optional (default is theme.defaultPalette)
  height: PropTypes.number,
  /**
   * Allows HighCharts chart update. Use it if you plan to provide new data.
   */
  allowUpdate: PropTypes.bool,
  // From withTheme
  theme: PropTypes.object.isRequired,
};

export default withTheme(PieChart);
