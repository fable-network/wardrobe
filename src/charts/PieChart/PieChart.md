A pie chart for a few data points and custom tooltip.
Example usage: theme distribution for an order.

```jsx
<PieChart
  title="Distribution of themes"
  data={require('../fixtures/pie-chart.js').data}
  tooltip={({ y, point: { percentage } }) => `Quantity: ${y}<br>Percent: ${percentage.toFixed(0)}%`}
/>
```

With custom colors:

```jsx
<PieChart
  title="Distribution of themes"
  data={require('../fixtures/pie-chart-custom-colors.js').data}
  colors={require('../fixtures/pie-chart-custom-colors.js').colors}
/>
```

With one data point and no title:

```jsx
<PieChart
  data={require('../fixtures/pie-chart-one.js').data}
  colors={require('../fixtures/pie-chart-one.js').colors}
  tooltip={({ y, point: { percentage } }) => `Quantity: ${y}<br>Percent: ${percentage.toFixed(0)}%`}
/>
```
