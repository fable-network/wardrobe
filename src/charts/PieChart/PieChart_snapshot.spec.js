/* eslint-disable */
// This file was automatically generated
import React from 'react';
import 'jest-styled-components';
import PieChart from './PieChart';


jest.mock('../../charts/HighChart', () => () => <high-charts />);

describe('PieChart Component', () => {
  describe('snapshots', () => {
    it('should match snapshot', () => {
    
      const wrapper1 = mount(
        <PieChart
          title="Distribution of themes"
          data={require('../fixtures/pie-chart.js').data}
          tooltip={({ y, point: { percentage } }) => `Quantity: ${y}<br>Percent: ${percentage.toFixed(0)}%`}
        />
      );
      expect(wrapper1).toMatchSnapshot();

      const wrapper2 = mount(
        <PieChart
          title="Distribution of themes"
          data={require('../fixtures/pie-chart-custom-colors.js').data}
          colors={require('../fixtures/pie-chart-custom-colors.js').colors}
        />
      );
      expect(wrapper2).toMatchSnapshot();

      const wrapper3 = mount(
        <PieChart
          data={require('../fixtures/pie-chart-one.js').data}
          colors={require('../fixtures/pie-chart-one.js').colors}
          tooltip={({ y, point: { percentage } }) => `Quantity: ${y}<br>Percent: ${percentage.toFixed(0)}%`}
        />
      );
      expect(wrapper3).toMatchSnapshot();

    });
  });
});
