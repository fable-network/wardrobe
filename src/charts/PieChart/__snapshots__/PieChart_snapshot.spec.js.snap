// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChart Component snapshots should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithTheme(PieChart)
    data={
      Array [
        Object {
          "name": "Christmas",
          "price": "EUR 5.282,90",
          "y": 10,
        },
        Object {
          "name": "New Year",
          "price": "EUR 5.282,90",
          "y": 20,
        },
        Object {
          "name": "Glitter",
          "price": "EUR 5.282,90",
          "y": 30,
        },
        Object {
          "name": "Hip Hop",
          "price": "EUR 5.282,90",
          "y": 40,
        },
        Object {
          "name": "Street",
          "price": "EUR 5.282,90",
          "y": 50,
        },
        Object {
          "name": "Sunny",
          "price": "EUR 5.282,90",
          "y": 60,
        },
        Object {
          "name": "Casual",
          "price": "EUR 5.282,90",
          "y": 70,
        },
      ]
    }
    title="Distribution of themes"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "allowUpdate": false,
      "data": Array [
        Object {
          "name": "Christmas",
          "price": "EUR 5.282,90",
          "y": 10,
        },
        Object {
          "name": "New Year",
          "price": "EUR 5.282,90",
          "y": 20,
        },
        Object {
          "name": "Glitter",
          "price": "EUR 5.282,90",
          "y": 30,
        },
        Object {
          "name": "Hip Hop",
          "price": "EUR 5.282,90",
          "y": 40,
        },
        Object {
          "name": "Street",
          "price": "EUR 5.282,90",
          "y": 50,
        },
        Object {
          "name": "Sunny",
          "price": "EUR 5.282,90",
          "y": 60,
        },
        Object {
          "name": "Casual",
          "price": "EUR 5.282,90",
          "y": 70,
        },
      ],
      "theme": undefined,
      "title": "Distribution of themes",
      "tooltip": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "allowUpdate": false,
        "data": Array [
          Object {
            "name": "Christmas",
            "price": "EUR 5.282,90",
            "y": 10,
          },
          Object {
            "name": "New Year",
            "price": "EUR 5.282,90",
            "y": 20,
          },
          Object {
            "name": "Glitter",
            "price": "EUR 5.282,90",
            "y": 30,
          },
          Object {
            "name": "Hip Hop",
            "price": "EUR 5.282,90",
            "y": 40,
          },
          Object {
            "name": "Street",
            "price": "EUR 5.282,90",
            "y": 50,
          },
          Object {
            "name": "Sunny",
            "price": "EUR 5.282,90",
            "y": 60,
          },
          Object {
            "name": "Casual",
            "price": "EUR 5.282,90",
            "y": 70,
          },
        ],
        "theme": undefined,
        "title": "Distribution of themes",
        "tooltip": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`PieChart Component snapshots should match snapshot 2`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithTheme(PieChart)
    colors={
      Array [
        "#d8d8d8",
        "#b1b1b1",
        "#7e7e7e",
        "#474747",
        "#1a1a1a",
      ]
    }
    data={
      Array [
        Object {
          "name": "Light Light Grey",
          "price": "EUR 5.282,90",
          "y": 10,
        },
        Object {
          "name": "Light Grey",
          "price": "EUR 5.282,90",
          "y": 20,
        },
        Object {
          "name": "Grey",
          "price": "EUR 5.282,90",
          "y": 30,
        },
        Object {
          "name": "Dark Grey",
          "price": "EUR 5.282,90",
          "y": 40,
        },
        Object {
          "name": "Dark Dark Grey",
          "price": "EUR 5.282,90",
          "y": 50,
        },
      ]
    }
    title="Distribution of themes"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "allowUpdate": false,
      "colors": Array [
        "#d8d8d8",
        "#b1b1b1",
        "#7e7e7e",
        "#474747",
        "#1a1a1a",
      ],
      "data": Array [
        Object {
          "name": "Light Light Grey",
          "price": "EUR 5.282,90",
          "y": 10,
        },
        Object {
          "name": "Light Grey",
          "price": "EUR 5.282,90",
          "y": 20,
        },
        Object {
          "name": "Grey",
          "price": "EUR 5.282,90",
          "y": 30,
        },
        Object {
          "name": "Dark Grey",
          "price": "EUR 5.282,90",
          "y": 40,
        },
        Object {
          "name": "Dark Dark Grey",
          "price": "EUR 5.282,90",
          "y": 50,
        },
      ],
      "theme": undefined,
      "title": "Distribution of themes",
      "tooltip": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "allowUpdate": false,
        "colors": Array [
          "#d8d8d8",
          "#b1b1b1",
          "#7e7e7e",
          "#474747",
          "#1a1a1a",
        ],
        "data": Array [
          Object {
            "name": "Light Light Grey",
            "price": "EUR 5.282,90",
            "y": 10,
          },
          Object {
            "name": "Light Grey",
            "price": "EUR 5.282,90",
            "y": 20,
          },
          Object {
            "name": "Grey",
            "price": "EUR 5.282,90",
            "y": 30,
          },
          Object {
            "name": "Dark Grey",
            "price": "EUR 5.282,90",
            "y": 40,
          },
          Object {
            "name": "Dark Dark Grey",
            "price": "EUR 5.282,90",
            "y": 50,
          },
        ],
        "theme": undefined,
        "title": "Distribution of themes",
        "tooltip": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`PieChart Component snapshots should match snapshot 3`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithTheme(PieChart)
    colors={
      Array [
        "#6F94AE",
      ]
    }
    data={
      Array [
        Object {
          "name": "Christmas",
          "price": "EUR 5.282,90",
          "y": 10,
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "allowUpdate": false,
      "colors": Array [
        "#6F94AE",
      ],
      "data": Array [
        Object {
          "name": "Christmas",
          "price": "EUR 5.282,90",
          "y": 10,
        },
      ],
      "theme": undefined,
      "tooltip": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "allowUpdate": false,
        "colors": Array [
          "#6F94AE",
        ],
        "data": Array [
          Object {
            "name": "Christmas",
            "price": "EUR 5.282,90",
            "y": 10,
          },
        ],
        "theme": undefined,
        "tooltip": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
